staging/src/k8s.io/apiserver/pkg/util/applypatch/applypatch.go

```diff
@@ -0,0 +1,257 @@
+package applypatch
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager"
+	"k8s.io/kube-openapi/pkg/util/proto"
+	prototesting "k8s.io/kube-openapi/pkg/util/proto/testing"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v4/merge"
+	"sigs.k8s.io/structured-merge-diff/v4/typed"
+)
+
+type TestFieldManager struct {
+	fieldManager *fieldmanager.FieldManager
+	apiVersion   string
+	emptyObj     runtime.Object
+	liveObj      runtime.Object
+}
+
+func (f *TestFieldManager) APIVersion() string {
+	return f.apiVersion
+}
+
+func (f *TestFieldManager) Reset() {
+	f.liveObj = f.emptyObj.DeepCopyObject()
+}
+
+func (f *TestFieldManager) Get() runtime.Object {
+	return f.liveObj.DeepCopyObject()
+}
+
+func (f *TestFieldManager) Apply(obj runtime.Object, manager string, force bool) error {
+	out, err := f.fieldManager.Apply(f.liveObj, obj, manager, force)
+	if err == nil {
+		f.liveObj = out
+	}
+	return err
+}
+
+func (f *TestFieldManager) Update(obj runtime.Object, manager string) error {
+	out, err := f.fieldManager.Update(f.liveObj, obj, manager)
+	if err == nil {
+		f.liveObj = out
+	}
+	return err
+}
+
+func (f *TestFieldManager) ManagedFields() []metav1.ManagedFieldsEntry {
+	accessor, err := meta.Accessor(f.liveObj)
+	if err != nil {
+		panic(fmt.Errorf("couldn't get accessor: %v", err))
+	}
+
+	return accessor.GetManagedFields()
+}
+
+func NewDefaultTestFieldManager(gvk schema.GroupVersionKind) TestFieldManager {
+	return NewTestFieldManager(gvk, "", nil)
+}
+
+func NewTestFieldManager(gvk schema.GroupVersionKind, subresource string, chainFieldManager func(fieldmanager.Manager) fieldmanager.Manager) TestFieldManager {
+	m := NewFakeOpenAPIModels()
```

Comment:
I don't think you'll want that, it makes things much harder to test. Not sure in which context you want to use that but I suspect we should have a way to provide a custom OpenAPI?

Reply:
To be clear, I don't think it's a bad thing to have an easy way to give the OpenAPI, but I think it should be plug and play, if I want to run this test code against CRD code, I should also be able to do so.

Reply:
Made changes to pass the openapi models as parameters, so user can pass models of their choice